### PR TITLE
metering and billing

### PR DESCRIPTION
**Pull request by yudong2015**
Init metering and billing module.

### CODE REVIEW

File: api/mbing.proto [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+		option (grpc.gateway.protoc_gen_swagger.options.openapiv2_operation) = {
+			summary: "create attribute"
+		};
+		option (google.api.http) = {
+			post: "/v1/mbing/attribute"
+			body: "*"
+		};
+	}
+
+	rpc ListAttribute (CommonListRequest) returns (ListAttributeResponse) {
```
[the remark this answers is not in the document]
Comment (yudong2015) [reply api/mbing.proto]: I have updated all abbreviation.

File: api/mbing.proto [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+	HKD = 1;
+	USD = 2;
+}
+
+//************************************* Sku *************************
+
+message AttributeName {
+	google.protobuf.StringValue attribute_name_id = 1;
+	google.protobuf.StringValue name = 2;
+	google.protobuf.StringValue description = 3;
```
Comment (chilianyi) [inline api/mbing.proto]: status
create_time
status_time

File: api/mbing.proto [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+	uint32 limit = 4;
+}
+
+message DescribeAttributeNamesResponse {
+	repeated AttributeName attribute_names = 1;
+}
+
+message AttributeUnit {
+	google.protobuf.StringValue attribute_unit_id = 1;
+	google.protobuf.StringValue name = 2;
```
Comment (chilianyi) [inline api/mbing.proto]: status
create_time
status_time

File: api/mbing.proto [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+
+message DescribeAttributeUnitsResponse {
+	repeated AttributeUnit attribute_units = 1;
+}
+
+message Attribute {
+	google.protobuf.StringValue attribute_id = 1;
+	google.protobuf.StringValue attribute_name_id = 2;
+	google.protobuf.StringValue attribute_unit_id = 3;
+	google.protobuf.StringValue value = 4;
```
Comment (chilianyi) [inline api/mbing.proto]: status
create_time
status_time

File: api/mbing.proto [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+}
+
+message CreateAttributeUnitResponse {
+	google.protobuf.StringValue attribute_unit_id = 1;
+}
+
+message DescribeAttributeUnitsRequest {
+	google.protobuf.StringValue attribute_unit_id = 1;
+	google.protobuf.StringValue name = 2;
+	uint32 offset = 3;
```
Comment (chilianyi) [inline api/mbing.proto]: status

File: api/mbing.proto [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+}
+
+message CreateAttributeNameResponse {
+	google.protobuf.StringValue attribute_name_id = 1;
+}
+
+message DescribeAttributeNamesRequest {
+	google.protobuf.StringValue attribute_name_id = 1;
+	google.protobuf.StringValue name = 2;
+	uint32 offset = 3;
```
Comment (chilianyi) [inline api/mbing.proto]: status

File: api/mbing.proto [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+}
+
+message CreateAttributeResponse {
+	google.protobuf.StringValue attribute_id = 1;
+}
+
+message DescribeAttributesRequest {
+	google.protobuf.StringValue attribute_id = 1;
+	google.protobuf.StringValue attribute_name_id = 2;
+	google.protobuf.StringValue attribute_unit_id = 3;
```
Comment (chilianyi) [inline api/mbing.proto]: status
google.protobuf.StringValue sort_key
google.protobuf.BoolValue reverse

File: api/mbing.proto [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+}
+
+message DescribeAttributesResponse {
+	repeated Attribute attributes = 1;
+}
+
+//SPU: standard product unit
+message Spu {
+	google.protobuf.StringValue spu_id = 1;
+	google.protobuf.StringValue resource_version_id = 2;
```
Comment (chilianyi) [inline api/mbing.proto]: version_id

File: api/mbing.proto [hunk 9]
```diff
@@ -0,0 +1,480 @@
...
+	repeated google.protobuf.StringValue attribute_name_ids = 3;
+	repeated google.protobuf.StringValue metering_attribute_name_ids = 4;
```
Comment (chilianyi) [inline api/mbing.proto]: repeated google.protobuf.StringValue  -> repeated string

need create_time

File: api/mbing.proto [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+}
+
+message CreateSpuRequest {
+	google.protobuf.StringValue resource_version_id = 1;
```
Comment (chilianyi) [inline api/mbing.proto]: change to version_id

File: api/mbing.proto [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+	repeated google.protobuf.StringValue attribute_name_ids = 2;
+	repeated google.protobuf.StringValue metering_attribute_name_ids = 3;
```
Comment (chilianyi) [inline api/mbing.proto]: google.protobuf.StringValue -> string

File: api/mbing.proto [hunk 12]
```diff
@@ -0,0 +1,480 @@
...
+message CreateSpuResponse {
+	google.protobuf.StringValue spu_id = 1;
+}
+
+//SKU: stock keeping unit
+message Sku {
+	google.protobuf.StringValue sku_id = 1;
+	google.protobuf.StringValue spu_id = 2;
+	repeated google.protobuf.StringValue attribute_ids = 3;
+	repeated google.protobuf.StringValue metering_attribute_ids = 4;
```
Comment (chilianyi) [inline api/mbing.proto]: oogle.protobuf.StringValue -> string
status
create_time
status_time

File: api/mbing.proto [hunk 13]
```diff
@@ -0,0 +1,480 @@
...
+}
+
+message CreateSkuRequest {
+	google.protobuf.StringValue spu_id = 1;
+	repeated google.protobuf.StringValue attribute_ids = 2;
```
Comment (chilianyi) [inline api/mbing.proto]: google.protobuf.StringValue -> string

File: api/mbing.proto [hunk 14]
```diff
@@ -0,0 +1,480 @@
...
+}
+
+message CreateSkuResponse {
+	google.protobuf.StringValue sku_id = 1;
+}
+
+message Price {
+    google.protobuf.StringValue price_id = 1;
+    google.protobuf.StringValue sku_id = 2;
+    google.protobuf.StringValue attribute_id = 3; //metering attribute id
```
Comment (chilianyi) [inline api/mbing.proto]: useless attribute_id

File: api/mbing.proto [hunk 15]
```diff
@@ -0,0 +1,480 @@
...
+	map<string, double> prices = 4;  //{upto: price}
+    Currency currency = 5;
```
Comment (chilianyi) [inline api/mbing.proto]: need status, create_time, status_time,  start_time, end_time

File: api/mbing.proto [hunk 16]
```diff
@@ -0,0 +1,480 @@
...
+	enum SkuTypes {
+		SKU = 0;
+		COMBATION = 1;
+		PROBATION = 2;
+	}
+	SkuTypes type = 5;
+}
+
+message MeteringRequest {
+	google.protobuf.StringValue resource_id = 1; //cluster id
```
Comment (chilianyi) [inline api/mbing.proto]: Useless comment

File: api/mbing.proto (hunk 7 above)
Comment (chilianyi) [reply api/mbing.proto]: type

File: api/mbing.proto [hunk 17]
```diff
@@ -0,0 +1,481 @@
...
+
+//************************************* Sku *************************
+
+message AttributeName {
+	google.protobuf.StringValue attribute_name_id = 1;
+	google.protobuf.StringValue name = 2;
+	google.protobuf.StringValue description = 3;
+}
+
+message CreateAttributeNameRequest {
```
Comment (chilianyi) [inline api/mbing.proto]: type

File: api/mbing.proto [hunk 18]
```diff
@@ -0,0 +1,481 @@
...
+	Currency currency = 6;
+}
+
+message CreateCombinationPriceRequest {
+	google.protobuf.StringValue combination_sku_id = 1;
+	google.protobuf.StringValue spu_id = 2;
+	google.protobuf.StringValue attribute_id = 3;
+	map<string, double> prices = 4;
+	Currency currency = 5;
+}
```
[the remark this answers is not in the document]
Comment (chilianyi) [reply api/mbing.proto]: A, B, C 三个产品  
Combination：
combination_sku_metering_attribute_id
sku_metering_attribute_id
message CombinationPrice {
     google.protobuf.StringValue combination_price_id = 1;
     google.protobuf.StringValue combination_sku_metering_attribute_id = 2;
 	map<string, Price> prices = 5;
 	Currency currency = 6;
 }

File: api/mbing.proto [hunk 19]
```diff
@@ -0,0 +1,481 @@
...
+
+message Limit {
+	enum Types {
+		RESOURCE = 0;
+		SKU = 1;
+		PRICE = 2;
+		USER = 3;
+		REGOIN = 4;
+	}
+	Types type = 1;
```
Comment (chilianyi) [inline api/mbing.proto]: No need type.

File: api/mbing.proto [hunk 20]
```diff
@@ -0,0 +1,481 @@
...
+	google.protobuf.DoubleValue value = 2;
+	enum UpdateValueTypes {
+		INCREMENTAL = 0;
+		REPLACE = 1;
+	}
+	UpdateValueTypes type = 3;
+}
+
+message MeteringSku {
+	google.protobuf.StringValue sku_id = 1;
```
Comment (chilianyi) [inline api/mbing.proto]: sku_metering_attribute_id

File: api/mbing.proto [hunk 21]
```diff
@@ -0,0 +1,481 @@
...
+//		option (grpc.gateway.protoc_gen_swagger.options.openapiv2_operation) = {
+//			summary: "update metering value about the sku_set"
+//		};
+//		option (google.api.http) = {
+//			put: "/v1/metering/skus"
+//			body: "*"
+//		};
+//	}
+//
+//	rpc CloseMetering (MeteringRequest) returns (MeteringResponse) {
```
Comment (chilianyi) [inline api/mbing.proto]: Terminate

**Comment by yudong2015** [comment]
Move metering and billing to another project.

**Pull request closed without merging**